File: android/app/src/main/java/deckers/thibault/aves/channelhandlers/MethodResultWrapper.java
package deckers.thibault.aves.channelhandlers;

import android.os.Handler;
import android.os.Looper;

import io.flutter.plugin.common.MethodChannel;

// ensure `result` methods are called on the main looper thread
public class MethodResultWrapper implements MethodChannel.Result {
    private MethodChannel.Result methodResult;
    private Handler handler;

    MethodResultWrapper(MethodChannel.Result result) {
        methodResult = result;
        handler = new Handler(Looper.getMainLooper());
    }

    @Override
    public void success(final Object result) {
        handler.post(() -> methodResult.success(result));
    }

    @Override
    public void error(final String errorCode, final String errorMessage, final Object errorDetails) {
        handler.post(() -> methodResult.error(errorCode, errorMessage, errorDetails));
    }

    @Override
    public void notImplemented() {
        handler.post(() -> methodResult.notImplemented());
    }
}

File: android/app/src/main/kotlin/deckers/thibault/aves/channel/calls/MethodResultWrapper.kt
package deckers.thibault.aves.channel.calls

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

// ensure `result` methods are called on the main looper thread
class MethodResultWrapper internal constructor(private val methodResult: MethodChannel.Result) : MethodChannel.Result {
    private val handler: Handler = Handler(Looper.getMainLooper())

    override fun success(result: Any?) {
        handler.post { methodResult.success(result) }
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        handler.post { methodResult.error(errorCode, errorMessage, errorDetails) }
    }

    override fun notImplemented() {
        handler.post { methodResult.notImplemented() }
    }
}

// ensure `result` methods are called on the main looper thread
class Coresult internal constructor(private val methodResult: MethodChannel.Result) : MethodChannel.Result {
    private val mainScope = CoroutineScope(Dispatchers.Main)

    override fun success(result: Any?) {
        mainScope.launch { methodResult.success(result) }
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        mainScope.launch { methodResult.error(errorCode, errorMessage, errorDetails) }
    }

    override fun notImplemented() {
        mainScope.launch { methodResult.notImplemented() }
    }
}
